### PR TITLE
Allow different shard scheme for input and output tensors in only Width Sharded Conv2d

### PR DESCRIPTION
**Pull request by sankarmanoj-tt**
### Ticket
#14123 

### Problem description
Having same shard config for input & output makes the op choose the largest common core count. Low core count leads to low perf and L1 OOM errors. 

### What's changed
Shard config for Input & Output Tensors can be independent for only Width Sharded Conv2d.

### Checklist
- [x] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/11727032596)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/test_new_conv2d.py [hunk 1]
```diff
@@ -436,13 +439,13 @@ def test_conv_features(
...
-    [ttnn.bfloat16, ttnn.bfloat8_b],
+    [ttnn.bfloat16],
 )
 @pytest.mark.parametrize(
     "activations_dtype",
-    [ttnn.bfloat16, ttnn.bfloat8_b],
+    [ttnn.bfloat16],
 )
-@pytest.mark.parametrize("auto_shard", [True], ids=["auto_shard"])
+@pytest.mark.parametrize("auto_shard", [False], ids=["no_auto_shard"])
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/test_new_conv2d.py:448]: why did you change auto shard again there should be both options
same questino for data formats?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 #include "common/constants.hpp"
 #include "impl/buffers/buffer_constants.hpp"
+#include "ttnn/operations/conv/conv2d/device/conv2d_op.hpp"
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:12]: I don't think we should have include for device ops here?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 3]
```diff
@@ -843,14 +849,24 @@ std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::T
         }
         conv_config.deallocate_activation = true;
     }
+    auto output_parallel_config = parallel_config;
+    if(conv_config.shard_layout == ttnn::TensorMemoryLayout::WIDTH_SHARDED)
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:853]: shard_layout is std::optional so it might not be present here

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 4]
```diff
@@ -843,14 +849,24 @@ std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::T
         }
         conv_config.deallocate_activation = true;
     }
+    auto output_parallel_config = parallel_config;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:852]: pls format the code

**Review by pavlejosipovic — COMMENTED** [review]
.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp [hunk 5]
```diff
@@ -148,11 +148,19 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_width_sharded_v2_impl(
     // TODO: Can conv_act_c_blocks be same as num_blocks_act_w?
     auto shard_shape = a.shard_spec().value().shape;
 
+    CoreRangeSet input_cores = a.memory_config().shard_spec.value().grid;
+    CoreRangeSet output_cores = output.memory_config().shard_spec.value().grid;
+    CoreRangeSet all_cores = output.memory_config().shard_spec.value().grid;
+    if(input_cores.num_cores()>output_cores.num_cores()){
+        all_cores = a.memory_config().shard_spec.value().grid;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:155]: a.memory_config().shard_spec.value().grid == input_cores rigth?
Pls format the code

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp [hunk 6]
```diff
@@ -54,7 +54,7 @@ Tensor optimized_conv_new(const Tensor& a, const Tensor &b, std::optional<const
     bool untilize_out, bool fuse_relu, MathFidelity math_fidelity,
     const OptimizedConvParallelizationConfig& parallelization_config,
     const OptimizedConvBlockConfig& block_config,
-    MemoryConfig memory_config,
+    MemoryConfig out_memory_config,
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:57]: do we have to do this rename, seems to add noise to the PR

**Review by pavlejosipovic — CHANGES_REQUESTED** [review]
> Descrition should state that this changes only affects width sharded convs

File: tests/ttnn/unit_tests/operations/test_new_conv2d.py (hunk 1 above)
Comment (sankarmanoj-tt) [reply tests/ttnn/unit_tests/operations/test_new_conv2d.py:448]: I had removed it during debugging. Will add back.

Comment (pavlejosipovic) [reply tests/ttnn/unit_tests/operations/test_new_conv2d.py:448]: but it's bad even on main, should be two options

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp [hunk 7]
```diff
@@ -229,18 +230,23 @@ void kernel_main() {
...
 
                 // wait on act semaphore value to become VALID (set by mcast sender after it multicasts data)
-
+                DPRINT<<"Core "<<this_core_id<<" waiting for act semaphore, sender "<<sender_x<<" "<<sender_y<<ENDL();
                 noc_semaphore_wait(act_mcast_receiver_semaphore_addr_ptr, VALID);
             }
-            cb_push_back(cb_id_act, act_block_num_tiles);
+
+            //Only output cores do compute
+            // if(this_core_id < num_output_cores)
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp:238]: Why is this commented out code here?

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp [hunk 8]
```diff
@@ -787,11 +791,11 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_width_sharded_v2_impl(
         );
 
         SetRuntimeArgs(program, act_kernel_id, CoreCoord(core_x, core_y), rt_args);
-
         SetRuntimeArgs(program, weights_kernel_id, CoreCoord(core_x, core_y), {
             core_index * weight_block_w_ntiles,
             b.buffer()->address(),
-            bias_base_address
+            bias_base_address,
+            (uint32_t) (core_index < output_num_cores ? 1 : 0)
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:798]: Why have `? 1 : 0`? Doesn't a bool map to uint32_t properly?

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp (hunk 6 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:57]: It would be better to be explicit about which tensor the memory config refers to. In the future, I would like to do something similar for block sharding along the width dimension.

File: tests/ttnn/unit_tests/operations/test_new_conv2d.py [hunk 9]
```diff
@@ -442,7 +445,7 @@ def test_conv_features(
     "activations_dtype",
     [ttnn.bfloat16, ttnn.bfloat8_b],
 )
-@pytest.mark.parametrize("auto_shard", [True], ids=["auto_shard"])
+@pytest.mark.parametrize("auto_shard", [True, False], ids=["auto_shard", "no_auto_shard"])
```
Comment (sankarmanoj-tt) [inline tests/ttnn/unit_tests/operations/test_new_conv2d.py:448]: @pavlejosipovic Enabled both auto_shard and no_auto_shard

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 4 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:851]: Done.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp (hunk 6 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:57]: I think we should leave it as `memory_config`, to be consistent with rest of the TTNN ops (we used to call it various things incl. `out_memory_config`, but then it was changed to `memory_config` in all ops)

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp [hunk 10]
```diff
@@ -148,11 +148,19 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_width_sharded_v2_impl(
     // TODO: Can conv_act_c_blocks be same as num_blocks_act_w?
     auto shard_shape = a.shard_spec().value().shape;
 
+    CoreRangeSet input_cores = a.memory_config().shard_spec.value().grid;
+    CoreRangeSet output_cores = output.memory_config().shard_spec.value().grid;
+    CoreRangeSet all_cores = output.memory_config().shard_spec.value().grid;
+    if(input_cores.num_cores()>output_cores.num_cores()) {
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:154]: Use proper formatting/spacing

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp [hunk 11]
```diff
@@ -259,8 +267,8 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_width_sharded_v2_impl(
     uint32_t num_blocks_act_w = act_matrix_width_ntiles / act_block_w_ntiles;
     uint32_t num_blocks_weight_w = weight_matrix_width_ntiles / weight_block_w_ntiles;
 
-    TT_FATAL(num_blocks_act_w%total_num_cores==0, "{} {}", num_blocks_act_w, total_num_cores);
-    uint32_t per_core_num_blocks_act_w = num_blocks_act_w/total_num_cores;
+    TT_FATAL(num_blocks_act_w%input_num_cores==0, "{} {}", num_blocks_act_w, input_num_cores);
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:270]: formatting

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp [hunk 12]
```diff
@@ -259,8 +267,8 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_width_sharded_v2_impl(
...
+    uint32_t per_core_num_blocks_act_w = num_blocks_act_w/input_num_cores;
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:271]: space around operators

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp [hunk 13]
```diff
@@ -307,9 +315,9 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_width_sharded_v2_impl(
     uint32_t weight_num_subblocks = weight_block_w_ntiles / out_subblock_w_ntiles;
     uint32_t weight_block_h_ntiles = act_block_w_ntiles;
     uint32_t weight_block_num_tiles = weight_block_w_ntiles * weight_block_h_ntiles;
-    uint32_t weight_block_in_channels_ntiles = input_channels_padded/(32*total_num_cores*per_core_num_blocks_act_w);
-    TT_FATAL(input_channels_padded>=(TILE_HEIGHT*total_num_cores), "Error");
-    TT_FATAL(input_channels_padded%(TILE_HEIGHT*total_num_cores)==0, "Error");
+    uint32_t weight_block_in_channels_ntiles = input_channels_padded/(32 * input_num_cores * per_core_num_blocks_act_w);
+    TT_FATAL(input_channels_padded>=(TILE_HEIGHT * input_num_cores), "Error");
+    TT_FATAL(input_channels_padded%(TILE_HEIGHT * input_num_cores)==0, "Error");
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:320]: Can you also update the error message to something meaningful?

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/weights_reader_width_sharded.cpp [hunk 14]
```diff
@@ -42,7 +44,10 @@ void kernel_main() {
     const uint32_t init_weight_start_tile_id = get_arg_val<uint32_t>(i); i+=1;
     const uint32_t weight_addr_dram_base = get_arg_val<uint32_t>(i); i+=1;
 
+
     uint32_t bias_addr_dram_base = get_arg_val<uint32_t>(i); i+=1;
+    const uint32_t is_active = get_arg_val<uint32_t>(i); i+=1;
+    // DPRINT<<"Active Weights : "<<is_active<<ENDL();
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/weights_reader_width_sharded.cpp:50]: remove the commented out code

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp (hunk 8 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp (hunk 7 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp]: Removed commented code.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp (hunk 10 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:154]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp (hunk 11 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:270]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp (hunk 12 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:271]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp (hunk 13 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp:320]: Done.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/weights_reader_width_sharded.cpp (hunk 14 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/weights_reader_width_sharded.cpp:50]: Done.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 2 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp]: Removed.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 3 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp [hunk 15]
```diff
@@ -3,6 +3,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include <stdint.h>
+#include <sys/param.h>
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp:6]: do we need this include?

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/activation_reader_width_sharded.cpp:6]: Co-pilot automatically adds header files. I didn't realize.  I will remove it.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 16]
```diff
@@ -835,14 +842,33 @@ Result conv2d(
         }
         conv_config.deallocate_activation = true;
     }
+
+    auto output_parallel_config = parallel_config;
+    if(!conv_config.shard_layout.has_value()) {
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:847]: this whole block lines 846-862 feels like it should belong in determine_parallel_config function()
is there any reason why it's not there?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 17]
```diff
@@ -985,8 +1011,8 @@ Result conv2d(
         if (memory_config.has_value() && memory_config.value() != conv_output.memory_config()) {
             conv_output = ttnn::to_memory_config(conv_output, memory_config.value(), std::nullopt);
         }
-
         return {conv_output, output_height, output_width, weight_tensor_on_device, bias_tensor_on_device};
+
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:1015]: nit: empty line

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 16 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:847]: I needed to add a new argument called output_parallel_config to the conv2d main function. I can put the logic that determines into into another function to clean up.

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:847]: conv_config also gets updated. I don't want to hide that in a different function.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 18]
```diff
@@ -835,14 +854,18 @@ Result conv2d(
         }
         conv_config.deallocate_activation = true;
     }
+
+    conv_config.shard_layout = parallel_config.shard_scheme;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:858]: can we remove this line now?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 19]
```diff
@@ -1048,50 +1070,6 @@ template std::tuple<ttnn::Shape, ttnn::MemoryConfig, bool, bool> get_conv_padded
     uint32_t in_channels,
     uint32_t out_channels);
 
-template std::tuple<ttnn::Tensor, ParallelConfig, bool, bool> shard_or_reshard_tensor_if_required<Device>(
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:1051]: yey!

**Review by pavlejosipovic — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]
